impl/maven-cli/src/main/java/org/apache/maven/cling/invoker/mvnsh/builtin/BuiltinShellCommandRegistryFactory.java

```diff
@@ -130,53 +129,133 @@ public String name() {
+                }
+            }
+        }
+
+        private void cd(CommandInput input) {
             try {
-                invoke(new CommandSession(), command, "--help");
-            } catch (Options.HelpException e) {
-                return compileCommandOptions(e.getMessage());
+                if (input.args().length == 1) {
+                    shellContext.cwd.change(input.args()[0]);
+                } else {
+                    shellContext.writer.accept("Error: 'cd' accepts only one argument");
+                }
             } catch (Exception e) {
-                // ignore
+                saveException(e);
+            }
+        }
+
+        private List<Completer> cdCompleter(String name) {
+            return List.of(new ArgumentCompleter(new Completers.DirectoriesCompleter(shellContext.cwd)));
+        }
+
+        private void ls(CommandInput input) {
+            try {
+                try (Stream<Path> list = Files.list(shellContext.cwd.get())) {
+                    list.forEach(file -> {
+                        if (Files.isDirectory(file)) {
+                            shellContext.writer.accept(file.getFileName().toString() + "/");
+                        } else {
+                            shellContext.writer.accept(file.getFileName().toString());
+                        }
+                    });
+                }
+            } catch (Exception e) {
+                saveException(e);
+            }
+        }
+
+        private void pwd(CommandInput input) {
+            try {
+                shellContext.writer.accept(shellContext.cwd.get().toString());
+            } catch (Exception e) {
+                saveException(e);
             }
-            return null;
         }
 
         private void mvn(CommandInput input) {
             try {
                 shellMavenInvoker.invoke(mavenParser.parseInvocation(
                         ParserRequest.mvn(input.args(), shellContext.invokerRequest.messageBuilderFactory())
+                                .cwd(shellContext.cwd.get())
                                 .build()));
             } catch (Exception e) {
                 saveException(e);
             }
         }
 
         private List<Completer> mvnCompleter(String name) {
-            List<Completer> completers = new ArrayList<>();
-            completers.add(new ArgumentCompleter(
-                    NullCompleter.INSTANCE,
-                    new Completers.OptionCompleter(
-                            new Completers.FilesCompleter(shellContext.invokerRequest::cwd), this::commandOptions, 1)));
-            return completers;
+            return List.of(new ArgumentCompleter(new StringsCompleter(
+                    "clean",
+                    "validate",
+                    "compile",
+                    "test",
+                    "package",
+                    "verify",
+                    "install",
+                    "deploy",
+                    "wrapper:wrapper")));
```

Comment:
Should that be computed somehow ?
If we can lookup a `LifecycleRegistry`, we could use it with something like:
```
context.lookup(LifecycleRegistry.class).stream().flatMap(Lifecycle:allPhases).map(Phase::name)
```

Reply:
Let me push the change.

Reply:
My intent here was to offer "most frequent" goals plus `wrapper:wrapper` as Maven4 does not have wrapper lifecycle anymore. But yes, make it complete by lookup is ok.